### ISSUE TITLE
Selecting a window in GUI mode

### ISSUE THREAD

**Issue by TheMegaTB** [lupoDharkael/flameshot#5]
When running in GUI selection mode it would be nice if the initial selection could be set to the window outlines by just left clicking a window without dragging.

**Comment by lupoDharkael** [comment]
A windows selection mode is great, but there is a problem with it:
I want to add support for Wayland, and the protocol doesn't support knowing where the windows are and its properties. Which makes adding such feature a very problematic thing to manage in such migration.

**Comment by Naatan** [comment]
Why not add it for X11 and then add it to Wayland once it becomes available? I don't see why applications should suffer because Wayland can't keep up. If anything stuff like this will urge Wayland to add support for it.

**Comment by lupoDharkael** [comment]
I'm thinking about implementing window detection using shape detection in the image itself. That way it wont depend on the desktop environment and could be easier to port Flameshot to other systems someday.
I'll try to find a self-contained solution for rectangle detection because I don't what to add a heavy dependency like OpenCV.

**Comment by JaneSmith** [comment]
I'd like a window screenshot feature that can work like this: https://askubuntu.com/questions/50604/application-for-taking-pretty-screenshots-like-os-x-does

**Comment by elypter** [comment]
just copy the way shutter does it.

**Comment by MurzNN** [comment]
Also you can lookup Hotshots code as example of window selector: https://github.com/obiwankennedy/HotShots

**Comment by ndac-todoroki** [comment]
> I'll try to find a self-contained solution for rectangle detection

This is certainly cool, but there are problems... if a window have transparent surroundings, and if you want to include that too, I'm afraid you can not detect the correct area to capture.

**Comment by lupoDharkael** [comment]
I agree, and now I think it's better to support native mechanics for window selection as a separated mode. 

I plan to release a new version soon™ when I end the exams.
This could be implemented for Flameshot 0.7 (I'd like to push more features but my time is limited)

**Comment by Metadorius** [comment]
great tool, the only thing I lack is the ability to screenshot a window. please make it available at least for x11!

**Comment by KiriKira** [comment]
Also need this feature. Waiting in hope!

**Comment by azzamsa** [comment]
Or just copy on how `maim` did it. `maim -i $(xdotool getactivewindow) filename.png`

**Comment by slondr** [comment]
Any update on this feature? Lacking window selection support is probably the only thing that makes me keep Shutter installed instead of just using Flameshot for everything.

**Comment by mserajnik** [comment]
Would also be very interested in this feature.

**Comment by jameswilliamknight** [comment]
I would love this feature!

**Comment by decryp2kanon** [comment]
i love this suggest! :+1:

**Comment by Massimo-B** [comment]
I agree my #799 could be solved by this like Left-Click to snapshot the window instead of a command line parameter. Just an option for including window decoration or not would be helpful.

**Comment by tiotrom** [comment]
This feature is so necessary :D

**Comment by Flimm** [comment]
For those looking for a workaround, in GNOME, you can press Alt-PrintScreen, and it will save a screenshot of the currently active window in your Pictures directory.

**Comment by tiotrom** [comment]
> For those looking for a workaround, in GNOME, you can press Alt-PrintScreen, and it will save a screenshot of the currently active window in your Pictures directory.

Oh awesome. This is so useful honestly! Thanks a lot. Had no idea

**Comment by MithicSpirit** [comment]
> I want to add support for Wayland, and the protocol doesn't support knowing where the windows are and its properties.

Has there been any change on this? Or are we still waiting on that being added to the protocol?

**Comment by klosteinmann** [comment]
Would highly appreciate GUI mode for Flameshot especially as windows user. That's a feature that I've been missing with Flameshot that has been available with Snip & Sketch / snipping bar.

**Comment by MithicSpirit** [comment]
I don't use Windows but the Linux version does have a GUI version (it can be started by either executing `flameshot gui` or through the system tray icon). That isn't exactly what this issue is discussing though so I think that if you can't start it maybe create a new issue.

**Comment by itanium21** [comment]
4 years later and still nothing? Only feature that I miss in FlameShot :/

**Comment by wael444** [comment]
is this a thing yet or am i blind?

**Comment by mmahmoudian** [comment]
@wael444 no it is not a thing yet. You can follow some if the discussions in #1814

If you want this now and you are on Linux and using X11, you can use this wrapper tool I wrote to do it:

https://codeberg.org/mehrad/dmenu_shot

Basically when you use the "Select_Window" it will allow you to click on one of your open windows and then it automatically handles the dimension of the selection.

**Comment by borgmanJeremy** [comment]
It's not going to happen unless qt adds a way to do this. It will be a nightmare to manage it across desktop environments and operating systems.

**Comment by mmahmoudian** [comment]
@aryadovoy if you want a quick way to do it with Flameshot in Linux on X11, you can use the Select_Window in this dmenu tool:
https://codeberg.org/mehrad/dmenu_shot

**Comment by CholoTook** [comment]
Vote

**Comment by WebDust21** [comment]
Just installed Flameshot...and found this issue less than 10 minutes later while looking for a way to do just this...

It's not impossible to do in Linux, as both GIMP and Spectacle can "capture active window" without an issue.

Yes, transparent edges are a minor issue--GIMP simply captures the entire transparent edge (including what's visible behind it).  We don't necessarily need a "perfect" solution, just a way to have Flameshot automatically select the active window (command line shortcut activated with Ctrl + PrntScrn, for example)

**Comment by borgmanJeremy** [comment]
It's not impossible on any platform but I am not going to merge something that is unmaintainable. We need a clear strategy for how to do this in a cross platform / cross desktop way.

**Comment by mmahmoudian** [comment]
@WebDust21 I 100% agree with @borgmanJeremy. If you are on X11 and want to do it, there are many easy ways to do so. Take a look at [this project of mine](https://codeberg.org/mehrad/dmenu_shot) in which you can select the window and it will tell Flameshot to focus on that area. You ca either use the software or grab that piece of code.

**Comment by aryadovoy** [comment]
@mmahmoudian, I see solution for all platforms — parsing image and cropping active window (mouse should be on this window to get position and determine which window should be parsed and cropped). Maybe it's too much, but it should work everywhere. :)

**Comment by mmahmoudian** [comment]
@aryadovoy then checkout #1814
Pull request would be highly appreciated if you have the time to work on it

**Comment by friday** [comment]
It could indeed be possible to use some kind of edge detection, but it would have to make some assumptions. You wouldn't be able to detect a black terminal window with a borderless theme on a black background for example.

**Comment by mmahmoudian** [comment]
@friday there is no image processing method in existence that can handle this case unless there are some other visual cues. The worst case scenario is having a borderless black terminal with opaque background which all its area is inside another slightly larger terminal. No machine learning method can dig that out unless it can get some extra information based on text or window decoration.

I think if we end up using any image processing method, we should not consider edge cases and we should just go with typical and normal cases. The ideal case would be that we can read window dimension and position info from Qt as it has no overhead on our side.

**Comment by XieJiSS** [comment]
This feature is quite needed anyway, despite the thread being inactive for nearly a year - should we pick this up again and consider how to implement it?

**Comment by mmahmoudian** [comment]
> should we pick this up again and consider how to implement it?

Sure, if you are up for it. PRs are always welcome. But I have to hide your comment as it does not add anything constructive to the thread and discussion.

**Comment by mmahmoudian** [comment]
> So this feature is still opening just because we MUST wait for Wayland?

No, it is open because we need someone to consider helping implementing it. If you are up for it, PRs are definitely welcome

**Comment by KES777** [comment]
Sadly, that we lack of this feature for 6 years.

**Comment by mmahmoudian** [comment]
@KES777 Please avoid cross posting. This in already very easy to do in X11 (have been possible for over a year). It is not possible on wayland though.

For X11 I have implemented this in a personal project under "selec window", and you can salvage the code:

https://codeberg.org/mehrad/dmenu_shot

**Comment by harold** [comment]
@mmahmoudian - thank you for this! I was able to adapt your code to a short script that works well for me under x11/i3:

```bash
#!/bin/bash
TMP_WINDOW_ID=$(xdotool selectwindow)
unset WINDOW X Y WIDTH HEIGHT SCREEN
eval $(xdotool getwindowgeometry --shell "${TMP_WINDOW_ID}")
xdotool windowfocus --sync "${TMP_WINDOW_ID}"
sleep 0.05
flameshot gui --region "${WIDTH}x${HEIGHT}+${X}+${Y}"
```

Ignore the haters, you're doing great! Thank you for all the efforts. :bow:

**Comment by RokeJulianLockhart** [comment]
We could ask the current window managers to separate their window detection code into a library or libraries, I suppose. Heck, we could just take theirs, do it for them, and then say "Hey, we consolidated all your code, want to use this and pool our development resources?" It seems more feasible than image detection, at least.

**Comment by Korb** [comment]
In the PicPick app, this feature is called "`Capture Window Control`", and it allows you to select not only the outer contours of any window, but also its internal areas. I propose to consider this function in this more universal way.

**Comment by grymmjack** [comment]
> @wael444 no it is not a thing yet. You can follow some if the discussions in #1814
> 
> If you want this now and you are on Linux and using X11, you can use this wrapper tool I wrote to do it:
> 
> https://codeberg.org/mehrad/dmenu_shot
> 
> Basically when you use the "Select_Window" it will allow you to click on one of your open windows and then it automatically handles the dimension of the selection.

I use display scaling on my KDE setup of 150% 

To accommodate this you need to add the scaling environment variables to get the correct geometry:
```sh
221c221
<         flameshot gui --region "${WIDTH}x${HEIGHT}+${X}+${Y}"
---
>         QT_AUTO_SCREEN_SCALE_FACTOR=1.5 QT_SCREEN_SCALE_FACTORS="" flameshot gui --region "${WIDTH}x${HEIGHT}+${X}+${Y}"
```
1.5 = 150% scaling, for example:
<img width="950" alt="image" src="https://github.com/user-attachments/assets/669318b2-49bf-4711-8248-bcf25964ee8c">

@mmahmoudian thank you for dmenu_shot! Also thank you for exposing dmenu! it's great.

I wonder, is there a way to get the screen scaling factor and automatically compensate that in your tool?

BTW, I found this information here in flameshot troubleshooting:
https://flameshot.org/docs/guide/troubleshooting/

**Comment by beedell-roke** [comment]
Note that this is more complex than merely defining a region to capture based upon the window geometry. Flameshot needs to isolate the window from the rest of the canvas, because it otherwise shall not capture areas that should be represented with the alpha channel due to non-square window decorations.

This is the pitfall that `PackageFamilyName: Microsoft.ScreenSketch_8wekyb3d8bbwe` ⪆ `Version: 11.2307.48.0` falls into, but which `spectacle-6.3.3-1` doesn't. You can see an example of this undermentioned, on its top left and right corners:

![Image](https://github.com/user-attachments/assets/55e7dbb7-619e-4465-b79a-f0ba7e93fd96)

Consequently, I suggest that instead of attempting to reinvent this, some examples from existing projects with this functionality be examined. ...if this isn't already what's occurring.

<p></p><details><p></p>

On that note, this is possible on AOSP already, too: [^1]

> ![Image](https://github.com/user-attachments/assets/45c18e76-191e-40ee-937a-ece21f5979f9)

That's not cropped.

[^1]: [`android.stackexchange.com/revisions/260313/5`](https://android.stackexchange.com/revisions/260313/5#:~:text=Select%20the%20navigation%20bar's%20%22Overview%22%20button.&text=Swipe%20to%20hover%20over%20the,Select%20the%20%22Screenshot%22%20button.)

<p></p></details><p></p>

**Comment by SpeeQz1** [comment]
@RokeJulianLockhart 

I believe you can find `gnome-screenshot` functionality for the capturing a window uitlity in here:

https://github.com/GNOME/gnome-screenshot/blob/master/src/screenshot-backend-x11.c

It can capture only the window and if for example the corners are rounded, it leaves the background there transparent instead of showing anything behind.

**Comment by beedell-roke** [comment]
@SpeeQz1, thanks. Forgot about GNOME Shell's: [^1] [^2]

<p></p><details><p></p>

<img width="1080" height="2340" alt="Image" src="https://github.com/user-attachments/assets/ff4fe5c6-2354-42fe-b076-993ad355300f" /> [^3]

<p></p></details><p></p>

For future reference, the relevant commit is [`GNOME/gnome-screenshot/blob/10a727d8287f1d47b36afbb8f601259a45b080ff/src/screenshot-backend-x11.c`](https://github.com/GNOME/gnome-screenshot/blob/10a727d8287f1d47b36afbb8f601259a45b080ff/src/screenshot-backend-x11.c#L1C1-L579C22:~:text=/-,screenshot%2Dbackend%2Dx11.c,-Latest%20commit) (since `HEAD`'s bound to eventually move that file).

Though, more equivalently, there's also [`gitlab.gnome.org/gnumdk/screenshot/-/blob/2fa85f97c420e34ef5df4251d642dabf02702344/src/screenshot-window.c`](https://gitlab.gnome.org/gnumdk/screenshot/-/blob/2fa85f97c420e34ef5df4251d642dabf02702344/src/screenshot-window.c#L21-361:~:text=History-,screenshot%2Dwindow.c,-9.69%20KiB), again for GNOME Desktop:

<p></p>

<img width="1380" height="1194" alt="Image" src="https://github.com/user-attachments/assets/d3d9c378-edd6-4287-a011-08e77955a9ff" /> [^3]

<p></p>

[^3]: [`forum.endeavouros.com/t/48795/7`](https://forum.endeavouros.com/t/how-to-actually-share-a-file-via-bluetooth/48795/7?u=rokejulianlockhart)

[^1]: [`discourse.gnome.org/t/30326/8`](https://discourse.gnome.org/t/how-to-capture-a-native-screencast-without-a-keyboard/30326/8?u=rokejulianlockhart)

[^2]: [`discussion.fedoraproject.org/t/1515/5`](https://discussion.fedoraproject.org/t/add-gnome-screenshot-on-fedora-oci/1515/5)
